File: metrics/src/cicd_script/report_devops_event.py
import os
import argparse
import json
import requests
from dotenv import load_dotenv

EVENT_TYPES = ['push', 'deploy', 'test pass']


class DevOpsEventReporter:

    def __init__(self, target_url: str, secret_token: str = "") -> None:
        self.target_url = target_url
        self.secret_token = secret_token

    def parse_cli_arguments(self) -> argparse.Namespace:

        parser = argparse.ArgumentParser()
        parser.add_argument('repo_name', help='The name of the repository')
        parser.add_argument('event', choices=EVENT_TYPES,
                            help=f'The event to log (must be one of: {EVENT_TYPES}')
        parser.add_argument(
            '--metadata', help='Additional JSON metadata about the event', default="")

        return parser.parse_args()

    def prepare_record(self, repo_name: str, event: str, metadata: str) -> dict:
        record = {
            'Repo': repo_name,
            'Event': event,
            'Metadata': metadata
        }

        return record

    def post_event(self, record: dict) -> requests.Response:

        headers = {
            'X-Secret-Token': self.secret_token,
            'Content-Type': 'application/json'
        }

        response = requests.post(self.target_url, json=record, headers=headers)

        return response


def main():

    load_dotenv()

    target_url = get_target_url()
    secret_token = get_secret_token()
    event_reporter = DevOpsEventReporter(target_url, secret_token)

    args = event_reporter.parse_cli_arguments()
    record = event_reporter.prepare_record(
        args.repo_name, args.event, args.metadata)
    response = event_reporter.post_event(record)

    if response.status_code == 200:
        print("Event logged successfully.")
    else:
        print(f"Event logging failed. {response.status_code} {response.text}")


def get_target_url() -> str:

    target_url = os.getenv('API_DEVOPS_EVENT_CATCHER', None)
    if target_url is None:
        raise ValueError(
            "API_DEVOPS_EVENT_CATCHER environment variable is not set.")

    return target_url


def get_secret_token() -> str:

    secret_token = os.getenv('DEVOPS_EVENTS_SECRET_TOKEN', None)
    if secret_token is None:
        raise ValueError(
            "DEVOPS_EVENTS_SECRET_TOKEN environment variable is not set.")

    return secret_token


if __name__ == "__main__":
    main()

File: metrics/src/cicd_script/report_devops_event.js
const axios = require('axios');
require('dotenv').config();

const EVENT_TYPES = ['push', 'deploy', 'test_pass', 'test_run'];

class DevOpsEventReporter {
    constructor(targetUrl, secretToken = "") {
        this.targetUrl = targetUrl;
        this.secretToken = secretToken;
    }

    parseCliArguments() {
        const args = process.argv.slice(2);
        const repoName = args[0];
        const event = args[1];
        const metadata = args[2] || "";

        if (!EVENT_TYPES.includes(event)) {
            throw new Error(`The event must be one of: ${EVENT_TYPES.join(', ')}`);
        }

        return { repoName, event, metadata };
    }

    prepareRecord(repoName, event, metadata) {
        return {
            Repo: repoName,
            Event: event,
            Metadata: metadata
        };
    }

    async postEvent(record) {
        const headers = {
            'X-Secret-Token': this.secretToken,
            'Content-Type': 'application/json'
        };
    
        try {
            const response = await axios.post(this.targetUrl, record, { headers });
            return response;
        } catch (error) {
            console.error("Error in postEvent:", error);
            return { status: error.response ? error.response.status : 500, data: error.response ? error.response.data : 'Internal Server Error' };
        }
    }
    
}

async function main() {
    const targetUrl = getTargetUrl();
    const secretToken = getSecretToken();
    const eventReporter = new DevOpsEventReporter(targetUrl, secretToken);

    const args = eventReporter.parseCliArguments();
    const record = eventReporter.prepareRecord(args.repoName, args.event, args.metadata);
    const response = await eventReporter.postEvent(record);

    if (response && response.status === 200) {
        console.log("Event logged successfully.");
    } else {
        console.log(`Event logging failed. Status: ${response ? response.status : 'N/A'}, Data: ${response ? response.data : 'N/A'}`);
    }
}


function getTargetUrl() {
    const targetUrl = process.env.API_DEVOPS_EVENT_CATCHER;
    if (!targetUrl) {
        throw new Error("API_DEVOPS_EVENT_CATCHER environment variable is not set.");
    }
    return targetUrl;
}

function getSecretToken() {
    const secretToken = process.env.DEVOPS_EVENTS_SECRET_TOKEN;
    if (!secretToken) {
        throw new Error("DEVOPS_EVENTS_SECRET_TOKEN environment variable is not set.");
    }
    return secretToken;
}

main();
